feat(ARCHITECTURE.md): add architectural overview (#1531)
mamba-org/rattler: ARCHITECTURE.md

# Rattler (monorepo) Architecture

> Rattler is a suite of Rust crates providing Conda functionality (indexing, solving, installing, etc.) without Python.

Reviewer: ```suggestion
Rattler is a suite of Rust crates providing Conda functionality (indexing, solving, installing, etc.).
```

Dont think Without python is relevant.

Its design reflects clear separation of concerns: each crate has a focused responsibility with well-defined interfaces.
This modularity enables reuse (e.g. Python/JS bindings reuse core crates), and isolates dependencies (e.g. a solve engine crate doesn’t depend on networking).
High-level tools (like the `rattler` CLI or *pixi*) orchestrate these crates: for example, solving and then installing an environment uses the `rattler_solve`, `rattler_index`, and `rattler` crates together.

> The **“Components”** section of the README lists all crates and their roles, guiding new contributors through the architecture.

Reviewer: This was removed in your PR.

Overall, Rattler’s architecture emphasizes **simplicity and orthogonality**: each crate does one job (e.g. *“download/stream packages”* or *“activate environments”*) so that the system is easier to understand, test, and extend.

> * **Core idea:** Provide Conda-like workflows from Rust (and other languages) with a clean API.

Reviewer: ```suggestion
* **Core idea:** Provide strong types, memory efficient data-structures and performant algorithms to interact with the conda package ecosystem from Rust (and other languages) with a clean API.
```

By centralizing these definitions, Rattler ensures consistency (everyone uses the same `Version` type, same repo format, etc.) and avoids duplication.
Design decisions include **ser/deser support** (so these types can be read/written from TOML/JSON) and **source attribution** (e.g. tracking which channel a package record came from).
Historically, this crate evolved from Rust parsers of Conda’s YAML metadata; it abstracts Conda’s naming and versioning conventions into type-safe Rust structs.
In summary, `rattler_conda_types` is the foundation of the codebase, capturing “what a Conda package/environment is” so that other crates can manipulate these concepts reliably.

## rattler\_package\_streaming

This crate handles downloading and unpacking Conda package archives.
It provides functionality to fetch a `.tar.bz2` or `.conda` file (from a URL or local cache), stream it through decompression, and expose its contents (files, metadata).
Key design points: it abstracts I/O so that higher layers just ask “download this package” without worrying about compression details.

> It uses robust Rust libraries (`fs-err`, streaming readers, etc.) to ensure safe file operations.

Reviewer: ```suggestion
```

Dont think this is relevant

> By separating package streaming, Rattler allows other parts (like `rattler_index` or `rattler_solve`) to use it without pulling in network or compression code.

Reviewer: ```suggestion
By separating package streaming, other parts (like `rattler_index` or `rattler_solve`) can deal with packages without having to worry about the internal workings of different package formats.
```

In practice, `rattler_package_streaming` is used by the install routines in the main `rattler` crate to fetch packages after solving.
This crate does not perform any dependency solving itself—it simply **handles package archives** (hence its name) and can be used independently wherever package files need processing.

## rattler\_repodata\_gateway

This crate is responsible for fetching and processing repodata (the *index* of available packages in a channel).

> It downloads `repodata.json` (and optionally legacy index formats), parses it, and provides APIs to query package metadata (names, versions, dependencies, etc.).

Reviewer: ```suggestion
It provides an API to acquire so called repodata while hiding the complex details from the user like index formats, which channels supports which format, exhaustive recursive search for package metadata, compression, deduplication and caching.
```

> The design isolates networking and data formats: it uses `rattler_networking` to fetch data and `rattler_conda_types` to represent the contents.
> It may also handle caching of repodata to avoid redundant downloads.

Reviewer: ```suggestion
```

Separating this crate lets the solver and install logic assume they already have the index loaded.
For example, during environment creation, Rattler will use `rattler_repodata_gateway` to load the channel indexes before solving.

> Historically, this reflects Conda’s own design: Repodata is distinct from package archives, so Rattler cleanly mirrors that by having a dedicated “gateway” for the repodata layer.

Reviewer: ```suggestion
```

## rattler\_shell

The `rattler_shell` crate encapsulates environment activation and command execution.
Its job is to modify a process’s environment (PATH, environment variables, etc.) and spawn subprocesses inside a given Conda environment prefix.
Design-wise, it knows about Conda’s activation scripts and directory layout, and it provides helpers to apply activation to the current process or a child process.
By extracting this logic into its own crate, Rattler ensures that creating an activated environment (e.g. “run Python from this env”) is a self-contained concern.
Other crates (like `rattler_menuinst`) also rely on `rattler_shell` to find prefixes or Python executables.
In practice, when the main `rattler` tool installs an environment, it will finally call into `rattler_shell` to set up the environment variables and launch the requested application.

> This clear separation reflects Conda’s concept of an “activated shell” and keeps OS-specific details (Windows vs Unix paths, etc.) contained.

Reviewer: ```suggestion
```

## rattler\_solve

The `rattler_solve` crate provides the dependency solver.
It is a backend-agnostic interface for solving the package satisfiability (SAT) problem: given requested package specs and a channel index, produce a list of concrete package versions to install.
Rattler’s solver wraps either the **libsolv** C library or the newer **resolvo** Rust solver.
Design goals include: no direct Conda/Python dependency (pure Rust), and flexibility to swap solver engines.
The API separates model (adding packages and constraints) from the solving step.
Other crates (and the `rattler` crate) call into `rattler_solve` when they need to figure out which packages satisfy a user’s request.
Its design is influenced by libsolv’s structures, but encapsulates them so that `rattler_solve` acts as a pure Rust library.
Because solving can be computationally intensive, it is a separate crate to avoid imposing solver dependencies on unrelated code.

> Documentation notes it is “backend agnostic”, reflecting this pluggable design.

Reviewer: ```suggestion
```

## rattler\_virtual\_packages

This crate detects system “virtual” packages (like `__unix`, `__linux`, `__glibc`, etc.) and represents them as packages for the solver.
Its job is to query the OS (CPU architecture, OS, glibc version, CPU features, etc.) and produce metadata so the solver knows what the system provides.
For example, if glibc 2.31 is present, it might add a virtual package `__glibc=2.31`.

> The design is a simple mapping of system queries to `rattler_conda_types::PackageRecord`s; it has no external dependencies beyond syscalls.

Reviewer: ```suggestion
```
